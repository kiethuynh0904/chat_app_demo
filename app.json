{
  "name": "chat_app_demo",
  "displayName": "chat_app_demo"
}