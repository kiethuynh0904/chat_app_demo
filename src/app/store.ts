import {configureStore, ThunkAction, Action} from '@reduxjs/toolkit';
import planetRoomReducer from '../app/slices/planetRoomSlice';

export const store = configureStore({
  reducer: {
    planetRoom: planetRoomReducer,
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
