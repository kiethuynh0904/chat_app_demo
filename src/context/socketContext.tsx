import React from 'react';
import io from 'socket.io-client';
// import {API_KEY} from '@env';

const SocketContext = React.createContext(null as any);

const SocketProvider = ({children}) => {
  const ENDPOINT = 'http://10.231.31.52:3000';
  const socket = io(ENDPOINT);
  return (
    <SocketContext.Provider value={socket}>{children}</SocketContext.Provider>
  );
};

export {SocketContext, SocketProvider};
