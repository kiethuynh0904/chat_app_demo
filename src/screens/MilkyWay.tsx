/* eslint-disable react-hooks/exhaustive-deps */
import React, {useContext, useEffect, useState} from 'react';
import {View, StyleSheet, FlatList, Alert} from 'react-native';

import {RouteProp, useRoute} from '@react-navigation/native';

import {Message, MessageInput} from '../components';
import {useAppDispatch, useAppSelector} from '../app/hook';
import {selectGroupMessage} from '../app/slices/planetRoomSlice';

import {Colors} from '../constants';
import {SocketContext} from '../context/socketContext';
import {Socket} from 'socket.io-client';

type ParamList = {
  Auth: {
    user: {
      name: string;
      id: string;
    };
  };
};

export const PLANET_ROOM = 'planetRoom'; // default room name

function MilkyWayScreen() {
  const route = useRoute<RouteProp<ParamList, 'Auth'>>();
  const {groupMessage} = useAppSelector(selectGroupMessage);

  const {user} = route.params;
  const socket: Socket = useContext(SocketContext);

  useEffect(() => {
    socket.emit('joinPlanetRoom', {user, room: PLANET_ROOM});

    return () => {
      console.log('leave room');
      socket.emit('leavePlanetRoom', {user, room: PLANET_ROOM});
    };
  }, []);

  return (
    <View style={styles.container}>
      <FlatList
        data={groupMessage}
        renderItem={({item}) => <Message user={user} message={item} />}
        inverted
        keyExtractor={item => item.id.toString()}
      />
      <MessageInput subscribe user={user} />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: Colors.white,
  },
});

export default MilkyWayScreen;
