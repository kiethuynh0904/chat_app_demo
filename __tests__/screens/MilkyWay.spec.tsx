import React from 'react';
import {render, waitFor} from '@testing-library/react-native';
import MilkyWay from '../../src/screens/MilkyWay';
import {store} from '../../src/app/store';
import {Provider} from 'react-redux';
import {SocketProvider} from '../../src/context/socketContext';
import * as hooks from '../../src/app/hook';

const mockData = [
  {
    id: 'm1',
    content: 'Hello Mars, now I can see you!',
    createdAt: 1652677931771,
    user: {
      id: 'earth',
      name: 'Earth',
    },
  },
  {
    id: 'm2',
    content: 'Oh! hi Earth, where are you from ?',
    createdAt: 1652677941993,
    user: {
      id: 'mars',
      name: 'Mars',
    },
  },
  {
    id: 'm3',
    content: 'I am in the same solar system as you?',
    createdAt: 1652677952156,
    user: {
      id: 'earth',
      name: 'Earth',
    },
  },
  {
    id: 'm4',
    content: 'Oh really! Im really happy to have new friends.',
    createdAt: 1652677969212,
    user: {
      id: 'mars',
      name: 'Mars',
    },
  },
];

let user = {
  name: 'Test',
  id: 'test',
};

jest.mock('@react-navigation/native', () => ({
  ...jest.requireActual('@react-navigation/native'),
  useNavigation: () => ({goBack: jest.fn()}),
  useRoute: () => ({
    params: {
      user,
    },
  }),
}));

describe('Testing milkyway screen', () => {
  test('test render flatlist', async () => {
    const screen = render(
      <Provider store={store}>
        <SocketProvider>
          <MilkyWay />
        </SocketProvider>
      </Provider>,
    );
    expect(screen).toMatchSnapshot();
  });

  test('test render item', async () => {
    jest
      .spyOn(hooks, 'useAppSelector')
      .mockReturnValue({groupMessage: mockData});

    const screen = render(
      <Provider store={store}>
        <SocketProvider>
          <MilkyWay />
        </SocketProvider>
      </Provider>,
    );
    const FlatlistComp = await waitFor(() => screen.getByTestId('messageList'));
    expect(FlatlistComp).toBeTruthy();
    expect(FlatlistComp.props.data.length).toBe(mockData.length);
  });
});
