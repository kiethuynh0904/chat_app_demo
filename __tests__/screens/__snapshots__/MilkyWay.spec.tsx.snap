// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing milkyway screen test render flatlist 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#fff",
      "flex": 1,
    }
  }
>
  <RCTScrollView
    data={Array []}
    getItem={[Function]}
    getItemCount={[Function]}
    invertStickyHeaders={true}
    inverted={true}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Array [
        Object {
          "transform": Array [
            Object {
              "scaleY": -1,
            },
          ],
        },
        undefined,
      ]
    }
    testID="messageList"
    viewabilityConfigCallbackPairs={Array []}
  >
    <View />
  </RCTScrollView>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "flexDirection": "row",
        },
        Object {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#f2f2f2",
          "borderColor": "#dedede",
          "borderRadius": 25,
          "borderWidth": 1,
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "center",
          "marginRight": 10,
          "padding": 8,
        }
      }
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          Array [
            Object {
              "color": undefined,
              "fontSize": 20,
            },
            Object {
              "marginHorizontal": 5,
            },
            Object {
              "fontFamily": "simple-line-icons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        
      </Text>
      <TextInput
        clearButtonMode="always"
        onChangeText={[Function]}
        placeholder="Aa"
        style={
          Object {
            "flex": 1,
          }
        }
        value=""
      />
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#3777f0",
          "borderRadius": 25,
          "height": 45,
          "justifyContent": "center",
          "width": 45,
        }
      }
      testID="sendMessageBtn"
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          Array [
            Object {
              "color": "#fff",
              "fontSize": 18,
            },
            undefined,
            Object {
              "fontFamily": "Ionicons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        
      </Text>
    </View>
  </View>
</View>
`;
